Unit11

crewcowl opened the pull request:
Вариант на проверку

crewcowl committed:
Save state

crewcowl committed:
unit11 v1

Bael commented on unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/service/QuestionServiceImpl.java:
```diff
@@ -25,20 +29,47 @@ public QuestionServiceImpl(QuestionRepository questionRepository,
 
-            answerRepository.save(answer);
-        }
+        answerRepository.findByQuestion(question).forEach(answerRepository::delete);
```
есть еще deleteAll

Bael commented on unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/service/SessionServiceImpl.java:
```diff
@@ -0,0 +1,150 @@
+
+    @Override
+    public double takePercent (double result,SessionEndDTO dto ) {
+        double percent = (result / dto.questionsList.size()) * 100;
```
может быть 0 элементов в списке вопросов?

Bael commented on unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/service/SessionServiceImpl.java:
```diff
@@ -0,0 +1,150 @@
+    }
+
+    @Override
+    public double getResultForSession (AnsweredQuestionDTO question, List<Answer> sessionAnswers) {
```
честно говоря расчет не очень понятен, могли бы описать ваш подход?

Bael commented on unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/service/SessionServiceImpl.java:
```diff
@@ -0,0 +1,150 @@
+        но если нет неправильных то мы в любом случае избегаем деления на 0, а если нет неправильных выбраных то эта функция все равно будет 0,
+        поэтому проверку оставил такой
+         */
+        double unCorrects = (unCorrectCheckedAnswers == 0) ? 0 : unCorrectCheckedAnswers/(allAnswers-correctAnswers);
```
что если все ответы корректные? такие вопросы часто встречаются в тестах. в задании мы специально оставили эту ошибку

crewcowl replied:
Как раз об этом выше в комментарии и пытался расписать

double unCorrects = (unCorrectCheckedAnswers == 0) ? 0 : unCorrectCheckedAnswers/(allAnswers-correctAnswers);
Собственно тут так как если все ответы корректны, то пользователь просто не сможет ткнуть не правильный ответ, и в unCorrects пойдет 0, а значит деления на 0 не произойдет

Bael replied:
да, теперь понял. Спасибо

Bael reviewed (changes requested):
Просьба поправить возможное деление на ноль в resultForManyCorrectAnswer 

Также алгоритм расчета процента правильных ответов у вас получился запутанный. По сути лучше было бы загрузить заранее все нужные данные для расчета оценки и подать их в класс\ функцию оценки. 
Такое и тестировать было бы легко. 
И изменять проще. 
И эффективнее вышло бы.

Bael reviewed (approved):
работа принята, поздравляю :)

crewcowl committed:
Исправлена функция подсчета для нескольких ответов
Исправлено удаление ответов при редактировании вопросов
Исправлен тест на проверку функции подсчета нескольких ответов
